lib/awful/rules.lua

```diff
@@ -204,23 +206,164 @@ function rules.matches_list(c, _rules)
+-- By default, Awesome provides 2 sources:
+--
+-- * `awful.rules`: A declarative matcher
+-- * `awful.spawn`: Launch clients with pre-defined properties
+--
+-- It is possible to register new callbacks to modify the properties table
+-- before it is applied. Each provider is executed sequentially and modifies the
+-- same table. If the first provider set a property, then the second can
+-- override it, then the third, etc. Once the providers are exhausted, the
+-- properties are applied on the client.
+--
+-- It is important to note that properties themselves have their own
+-- dependencies. For example, a `tag` property implies a `screen`. Therefor, if
+-- a `screen` is already specified, then it will be ignored when the rule is
+-- executed. Properties also have their own priorities. For example, the
+-- `titlebar` and `border_width` need to be applied before the `x` and `y`
+-- positions are set. Otherwise, it will be off or the client will shift
+-- upward everytime Awesome is restarted. A rule source *cannot* change this.
+-- It is up to the callback to be aware of the dependencies and avoid to
+-- introduce issues. For example, if the source wants to set a `screen`, it has
+-- to check if the `tag`, `tags` or `new_tag` are on that `screen` or remove
+-- those properties. Otherwise, they will be ignored once the rule is applied.
+--
+-- @tparam string name The provider name. It must be unique.
+-- @tparam function callback The callback that is called to produce properties.
+-- @tparam client callback.c The client
+-- @tparam table callback.properties The current properties. The callback should
+--  add to and overwrite properties in this table
+-- @tparam table callback.callbacks A table of all callbacks scheduled to be
+--  executed after the main properties are applied.
+-- @tparam[opt={}] table run_after A list of names of sources this source depends on.
+-- @tparam[opt={}] table run_before A list of names of sources this source have a
+--  priority over.
+-- @treturn boolean Returns false if a dependency conflict was found.
+function rules.add_rule_source(name, callback, run_after, run_before)
+    local min_index, max_index = 1, #rules.rule_sources+1
+
+    run_after  = run_after  or {}
+    run_before = run_before or {}
+    assert(type(run_after) == "table")
+    assert(type(run_before) == "table")
+
+    for k, v in ipairs(rules.rule_sources) do
+        -- Names must be unique
+        assert(
+            v.name ~= name,
+            "Name must be unique, but '" .. name .. "' was already registered."
+        )
+
+        for _, v2 in ipairs(run_after) do
+            if v.name == v2 then
+                min_index = math.max(min_index, k+1)
             end
         end
+
+        for _, v2 in ipairs(run_before) do
+            if v.name == v2 then
+                max_index = math.min(max_index, k)
+            end
+        end
+    end
+
+    if max_index + 1 < min_index then return false end
```

Comment:
> It's even worst for number ordering. priority 1 (a) + priority 1 (b) and you can't insert c between a and b. 

I'd say that `awful.rules` gets priority (or value or whatever) 0 and `awful.spawn` gets 10000. That should leave a lot of space so that people can pick non-colliding numbers where necessary. This does not solve all problems, as you point out, but at least it should solve all issues that depend on the order in which the rule sources are added.